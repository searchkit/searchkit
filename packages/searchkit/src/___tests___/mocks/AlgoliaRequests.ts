export const SimpleRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['type:movie']],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: 'shawshank',
      tagFilters: ''
    }
  }
]

export const nonDynamicFacetRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facets: ['type', 'actors', 'imdbrating'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const nonDynamicFacetRequestOneFilter = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['actors:Alice']],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const DisjunctiveExampleRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['type:movie']],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: 'shawshank',
      tagFilters: ''
    }
  },
  {
    indexName: 'imdb_movies',
    params: {
      analytics: false,
      attributesToHighlight: [],
      attributesToRetrieve: [],
      attributesToSnippet: [],
      clickAnalytics: false,
      facets: 'type',
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      hitsPerPage: 1,
      maxValuesPerFacet: 10,
      page: 0,
      query: 'shawshank',
      tagFilters: ''
    }
  }
]

export const NestedQueryRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['user.first:Alice']],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  },
  {
    indexName: 'imdb_movies',
    params: {
      analytics: false,
      attributesToHighlight: [],
      attributesToRetrieve: [],
      attributesToSnippet: [],
      clickAnalytics: false,
      facets: 'user.first',
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      hitsPerPage: 1,
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const TwoFiltersNestedQueryRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['user.first:Alice', 'user.first:John']],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  },
  {
    indexName: 'imdb_movies',
    params: {
      analytics: false,
      attributesToHighlight: [],
      attributesToRetrieve: [],
      attributesToSnippet: [],
      clickAnalytics: false,
      facets: 'user.first',
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      hitsPerPage: 1,
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const NumericFilterNestedQueryRequest = [
  {
    indexName: 'my-index-000001',
    params: {
      facetFilters: [],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      numericFilters: ['user.price>=90'],
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const FacetQueryExampleRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetName: 'type',
      facetQuery: 'm',
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxFacetHits: 10,
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const NumericFiltersExampleRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facets: ['type', 'actors', 'imdbrating'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      numericFilters: ['imdbrating>=90'],
      page: 0,
      query: '',
      tagFilters: ''
    }
  },
  {
    indexName: 'imdb_movies',
    params: {
      analytics: false,
      clickAnalytics: false,
      facets: 'imdbrating',
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      hitsPerPage: 0,
      maxValuesPerFacet: 10,
      page: 0,
      query: ''
    }
  }
]

export const NestedFacetQueryExampleRequest = [
  {
    indexName: 'my-index-000001',
    params: {
      facetName: 'user.first',
      facetQuery: 'har',
      facets: ['user.first', 'user.price'],
      highlightPostTag: '</ais-highlight-0000000000>',
      highlightPreTag: '<ais-highlight-0000000000>',
      maxFacetHits: 10,
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]
