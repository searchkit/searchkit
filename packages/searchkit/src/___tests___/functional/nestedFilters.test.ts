import Client, { AlgoliaMultipleQueriesQuery } from '../../'
import nock from 'nock'
import { NestedQueryRequest } from '../mocks/AlgoliaRequests'
import { ExampleNestedFiltersResponse } from '../mocks/ElasticsearchResponses'

describe('Nested Filters', () => {
  it('Numeric nested filter with 2 numeric filters', async () => {
    const client = new Client({
      connection: {
        host: 'http://localhost:9200',
        apiKey: 'a2Rha1VJTUJMcGU4ajA3Tm9fZ0Y6MjAzX2pLbURTXy1hNm9SUGZGRlhJdw=='
      },
      search_settings: {
        highlight_attributes: ['title'],
        search_attributes: ['title'],
        result_attributes: ['title'],
        facet_attributes: [
          {
            attribute: 'user.price',
            field: 'price',
            type: 'numeric',
            nestedPath: 'user'
          }
        ]
      }
    })

    nock('http://localhost:9200')
      .post('/_msearch', (requestBody: any) => {
        expect(requestBody).toMatchSnapshot('ES Request')
        const x = JSON.parse(requestBody.split('\n')[1])

        return true
      })
      .reply(200, {
        ...ExampleNestedFiltersResponse,
        responses: [ExampleNestedFiltersResponse.responses[0]]
      })

    const response = await client.handleInstantSearchRequests([
      {
        ...NestedQueryRequest[0],
        params: {
          ...NestedQueryRequest[0].params,
          numericFilters: ['user.price>=40', 'user.price<=140'],
          facetFilters: []
        }
      }
    ] as AlgoliaMultipleQueriesQuery[])

    expect(response).toMatchSnapshot()
  })
})
