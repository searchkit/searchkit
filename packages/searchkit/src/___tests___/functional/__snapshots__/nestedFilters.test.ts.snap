// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nested Filters Numeric nested filter with 2 numeric filters 1`] = `
{
  "results": [
    {
      "appliedRules": [],
      "exhaustive": {
        "facetsCount": true,
        "nbHits": true,
        "typo": true,
      },
      "exhaustiveFacetsCount": true,
      "exhaustiveNbHits": true,
      "exhaustiveTypo": true,
      "facets": {
        "user.first": {
          "Alice": 1,
          "John": 2,
          "Smith": 1,
        },
      },
      "facets_stats": {},
      "hits": [
        {
          "_highlightResult": {},
          "objectID": "1",
        },
        {
          "_highlightResult": {},
          "objectID": "2",
        },
      ],
      "hitsPerPage": 20,
      "index": "imdb_movies",
      "nbHits": 2,
      "nbPages": 1,
      "page": 0,
      "params": "facetFilters=&facets=*&highlightPostTag=%3C%2Fem%3E&highlightPreTag=%3Cem%3E&maxValuesPerFacet=10&page=0&query=&tagFilters=&numericFilters=user.price%3E%3D40%2Cuser.price%3C%3D140",
      "processingTimeMS": 1,
      "query": "",
      "renderingContent": {
        "facetOrdering": {
          "facets": {
            "order": [
              "user.price",
            ],
          },
          "values": {
            "user.price": {
              "sortRemainingBy": "count",
            },
          },
        },
      },
    },
  ],
}
`;

exports[`Nested Filters Numeric nested filter with 2 numeric filters: ES Request 1`] = `
"{"index":"imdb_movies"}
{"aggs":{"user.":{"nested":{"path":"user"},"aggs":{"user.price$_stats":{"stats":{"field":"user.price"}},"user.price$_entries":{"terms":{"field":"user.price","size":10}}}}},"query":{"bool":{"filter":[{"nested":{"path":"user","inner_hits":{},"query":{"bool":{"filter":[{"range":{"user.price":{"gte":"40"}}},{"range":{"user.price":{"lte":"140"}}}]}}}}],"must":{"match_all":{}}}},"size":20,"from":0,"_source":{"includes":["title"]},"highlight":{"pre_tags":["<em>"],"post_tags":["</em>"],"fields":{"title":{"number_of_fragments":0}}}}
"
`;
